### PR TITLE
api & test cleanups for platform differences

### PR DESCRIPTION
**Pull request by chrisvanheuveln**
1. Platform excludes in the yaml file to handle N7K vs N77, mt_full vs mt_lite, etc
2. Related to the first item, these are mostly feature enablement and feature checking changes
3. Test file changes: we had a lot of tests that we couldn't run on N3/N7/N9 without proper setup and teardowns (e.g. the vxlan, vni, vlan tests that rely on mt_full need a special line card to be present, whereas the mt_lite properties don't need the card but only work on N3/N9.

Test results from affected tests:

```

---
test_interface.rb
N7: 44 runs, 44 assertions, 10 failures, 14 errors, 1 skips (failures unrelated to my changes)
N9: 44 runs, 320 assertions, 0 failures, 0 errors, 2 skips
N3: 44 runs, 343 assertions, 0 failures, 0 errors, 5 skips

---
test_interface_service.rb
N7:  4 runs, 23 assertions, 0 failures, 0 errors, 0 skips
N39: 4 runs, 0 assertions, 0 failures, 0 errors, 4 skips

---
test_interface_svi.rb
N7:
N39: 19 runs, 46 assertions, 0 failures, 0 errors, 0 skips

---
test_interface_switchport.rb
N7: 53 runs, 59 assertions, 5 failures, 6 errors, 0 skips (failures unrelated to my changes)
N39:53 runs, 97 assertions, 0 failures, 0 errors, 0 skips

---
test_platform.rb
N7:  14 runs, 16 assertions, 3 failures, 0 errors, 1 skips (failures unrelated to my changes)
N39: 14 runs, 19 assertions, 0 failures, 0 errors, 0 skips

---
test_vdc.rb
N7: 3 runs, 6 assertions, 0 failures, 0 errors, 0 skips
N39:3 runs, 0 assertions, 0 failures, 0 errors, 3 skips

---
test_vni.rb
N7:  3 runs, 7 assertions, 0 failures, 0 errors, 1 skips
N39: 3 runs, 8 assertions, 0 failures, 0 errors, 2 skips

---
test_vlan.rb
N7: 21 runs, 44 assertions, 0 failures, 3 errors, 0 skips (failures unrelated to my changes)
N9: 21 runs, 45 assertions, 0 failures, 0 errors, 1 skips
N3: 21 runs, 45 assertions, 0 failures, 0 errors, 1 skips

---
test_vtp.rb
N7:  31 runs, 123 assertions, 2 failures, 0 errors, 0 skips (failures unrelated to my changes)
N39: 31 runs, 123 assertions, 0 failures, 0 errors, 0 skips

---
test_vrf.rb
N7: 8 runs, 17 assertions, 0 failures, 1 errors, 1 skips (failures unrelated to my changes)
N9: 8 runs, 21 assertions, 0 failures, 0 errors, 0 skips
N3: 8 runs, 19 assertions, 0 failures, 0 errors, 1 skips

---
test_vxlan_vtep.rb
N7: 7 runs, 54 assertions, 0 failures, 0 errors, 0 skips
N9: 7 runs, 22 assertions, 0 failures, 0 errors, 1 skips
N3: 7 runs, 19 assertions, 0 failures, 0 errors, 2 skips

---
test_router_bgp.rb
N7: 62 runs, 227 assertions, 1 failures, 0 errors, 2 skips
N9: 62 runs, 235 assertions, 0 failures, 0 errors, 0 skips
N3: 62 runs, 231 assertions, 0 failures, 1 errors, 0 skips (failures unrelated to my changes)

---
test_bgp_neighbor.rb
N7: 22 runs, 206 assertions, 0 failures, 1 errors, 0 skips (failures unrelated to my changes)
N39: 22 runs, 220 assertions, 0 failures, 0 errors, 0 skips

---
```

### CODE REVIEW

File: lib/cisco_node_utils/cmd_ref/vni.yaml [hunk 1]
```diff
@@ -61,6 +61,24 @@ mapped_vlan:
...
+mt_full_support:
+  # This is only used for determining support for Multi-Tenancy Full
+  kind: boolean
+  /N7/:
+    default_only: true
+  else:
+    # this feature is always off on these platforms and cannot be changed
+    default_only: false
+
+mt_lite_support:
```
Comment (alok-aggarwal) [inline lib/cisco_node_utils/cmd_ref/vni.yaml:73]: what about 5k and 6k ? Are they mt full or lite?

File: lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml [hunk 2]
```diff
@@ -10,9 +10,28 @@ all_interfaces:
   config_get_token: '/^interface (.*)$/'
 
 host_reachability:
-  config_get_token_append: '/^host-reachability protocol (\S+)/'
-  config_set_append: '<state> host-reachability protocol <proto>'
-  default_value: 'flood'
+  /N(7|9)/:
+    config_get_token_append: '/^host-reachability protocol (\S+)/'
```
Comment (alok-aggarwal) [inline lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml:14]: Is host-reachability not available in 3k/5k/6k?

File: tests/test_vni.rb [hunk 3]
```diff
@@ -21,22 +22,48 @@
...
+    v = Vdc.new('default')
+    v.limit_resource_module_type = 'f3' unless
+      v.limit_resource_module_type == 'f3'
+    config('no feature vni')
+    config('no feature nv overlay')
+  end
+
+  def mt_lite_env_setup
+    skip('Platform does not support MT-lite') unless Vni.mt_lite_support
+    config('no feature vni')
```
Comment (alok-aggarwal) [inline tests/test_vni.rb]: There is no "feature vni" on n9k (mt-lite)

File: lib/cisco_node_utils/cmd_ref/vni.yaml (hunk 1 above)
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/vni.yaml:73]: You tell me :-)

I thought 5/6k were neither at the moment but getting support for mt full at some point?

File: lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml (hunk 2 above)
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml:14]: It is not available on my n3k. Again I'll need your input for 5/6k support expectations.

File: tests/test_vni.rb (hunk 3 above)
Comment (chrisvanheuveln) [inline tests/test_vni.rb]: fixed

File: lib/cisco_node_utils/cmd_ref/vni.yaml (hunk 1 above)
Comment (alok-aggarwal) [inline lib/cisco_node_utils/cmd_ref/vni.yaml:73]: Just checked with my team, 5k/6k are mt-lite, so you may wanna add them under mt-lite here.

File: lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml (hunk 2 above)
Comment (alok-aggarwal) [inline lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml:14]: looks okay for now. We can tweak it when we test throughly for 5/6k

File: lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml [hunk 4]
```diff
@@ -10,9 +10,28 @@ all_interfaces:
...
+  /N(7)/:
+    default_only: true
+  else:
+    # this feature is always off on these platforms and cannot be changed
+    default_only: false
+
+mt_lite_support:
+  # This is only used for determining support for Multi-Tenancy Lite
+  kind: boolean
+  /N(3|9)/:
```
Comment (alok-aggarwal) [inline lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml:30]: Please add 5k, 6k under mt-lite

File: lib/cisco_node_utils/cmd_ref/vni.yaml (hunk 1 above)
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/vni.yaml:73]: My n56128P switch does not seem to support it:

```
n8k-136(config)# feature nv overlay
Error: feature nv overlay failed dependencies not satisfied yet

n8k-136(config)# feature vn-segment-vlan-based 
Error: vnseg_vlan control, vn-segment depends on feature-set fabricpath or feature nv overlay (0x411000bd)
```

If this is a license issue, how do I get around it? ie. how to check the current license; and how to fix it?

Comment (alok-aggarwal) [inline lib/cisco_node_utils/cmd_ref/vni.yaml:73]: Here's the license guide:
http://www.cisco.com/c/en/us/td/docs/switches/datacenter/sw/nx-os/licensing/guide/b_Cisco_NX-OS_Licensing_Guide/b_Cisco_NX-OS_Licensing_Guide_chapter_01.html

For now though, I suggest to add a story to backlog for adding mt-lite puppet support for 5/6k. We can look at it when we test 5/6k

Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/vni.yaml:73]: Agreed. 
https://rally1.rallydev.com/#/44221966144d/detail/userstory/49030571629

File: lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml (hunk 4 above)
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/vxlan_vtep.yaml:30]: Defer to US 62787.

**Pull request merged**